doc/ExternalDependencies.md

```diff
@@ -0,0 +1,255 @@
+  dependency.
+
+These goals necessitate a few design decisions. For example, as crates are the
+smallest unit of compilation in Rust, external dependencies must be included
+through new crates added to the Tock source tree so they can be individually
+included or excluded in specific builds. Also, crates provide a namespace to use
+to identify when external dependencies are being incorporated.
+
+Additionally, we avoid using traits or HIL-like interfaces for dependencies
+(i.e. core Tock capsules/modules would use a Tock-defined trait much like
+capsules use HILs, and a wrapper would use the external dependency to implement
+the trait) to avoid the overhead of implementing and maintaining a wrapper to
+implement the trait. While architecturally this has advantages, the overhead was
+deemed too burdensome for the expected benefit.
+
+We explicitly document the goals to help motivate the specific design in the
+remainder of this document. Also, this policy may change in the future, but
+these goals should be considered in any future updates.
+
+### Dependency Structure of Tock-Internal Crates
+
+Following from the above, an external dependency added to a crate which is
+depended on internally within Tock (e.g. the `kernel` crate) will have a higher
+impact than a dependency added to a crate with no reverse dependencies (e.g. a
+board crate). Thus, this policy is increasingly liberal with crate-types that
+have fewer reverse dependencies.
+
+This document considers Tock's crate structure by referring to the following
+types of crates internal to Tock:
+
+- the kernel crate: `kernel/`
+- arch crates: crates in the `arch/` directory
+- chip crates: crates in the `chips/` directory
+- board crates: crates in the `boards/` directory
+- capsule crates: crates in the `capsules/` directory
+
+Furthermore, this policy assumes the following rules regarding crate
+dependencies internal to Tock:
+
+- a _board crate_ is not a dependency of any other Tock-internal crate
+- a _chip crate_ is only a dependency of _board crates_ or other
+  _chip crates_
+- a _capsule crate_ is only a dependency of other _capsule crates_ or _board
+  crates_
+- an _arch crate_ may only depend on the _kernel crate_ and other
+  _arch crates_
+- the _kernel crate_ does not depend on _arch_, _chip_, _board_, or _capsule
+  crates_
+
+## External Dependency Selection
+
+External dependencies can be added to Tock on a case-by-case basis. Each
+dependency will be reviewed for inclusion, according to the criteria in this
+section. The requirements are intentionally strict.
+
+There are two general methods to for including an external dependency in the
+Tock kernel: core external dependencies, and capsule- or board-specific
+external dependencies.
+
+- Core external dependencies may be used in "core" Tock crates, such as the
+  _kernel_, _arch_, and _chip crates_. Notably, individual boards generally do
+  not have a reasonable choice whether to depend on those crates.
```

Comment:
Why don't we just remove this?

I think the two categories are "boards" and "not boards".

Reply:
The "core" notation is I think a leftover from the earlier version of the draft, and I'm not sure there is a good reason to keep it.

Reply:
> The "core" notation is I think a leftover from the earlier version of the draft, and I'm not sure there is a good reason to keep it.

I agree, and I think this would make the document much less confusing overall. Along with that I would remove the policies around _core external dependencies_ overall. We can punt on that until we need it, and resurrect the text. I just didn't want to delete it without discussing first.

> I think the two categories are "boards" and "not boards".

I think we do now have "boards", "capsules" and the rest, but the document should reflect that properly now.

Reply:
Okay, I have removed the notion of "core external dependencies" and changed the strict requirements on external dependencies (maturity, small dependency tree, etc.) to be guidelines for dependency selection. Let me know what you think!